rest_framework_json_api/filters/filter.py

```diff
@@ -0,0 +1,135 @@
+        `filter[<something>]` to comply with the jsonapi spec requirement to use the filter
+        keyword. The default is "search" unless overriden but it's used here just to make sure
+        we don't complain about it being an invalid filter.
+        """
+        # TODO: find a better way to deal with search_param.
+        search_param = api_settings.SEARCH_PARAM
+
+        # Make this regex check for 'filter' as well as 'filter[...]'
+        # Leave other incorrect usages of 'filter' to JSONAPIQueryValidationFilter.
+        # See http://jsonapi.org/format/#document-member-names for allowed characters
+        # and http://jsonapi.org/format/#document-member-names-reserved-characters for reserved
+        # characters (for use in paths, lists or as delimiters).
+        # regex `\w` matches [a-zA-Z0-9_].
+        # TODO: U+0080 and above allowed but not recommended. Leave them out for now. Fix later?
+        #       Also, ' ' (space) is allowed within a member name but not recommended.
+        filter_regex = re.compile(r'^filter(?P<ldelim>\[?)(?P<assoc>[\w\.\-]*)(?P<rdelim>\]?$)')
+
+        def validate_filter(self, keys, filterset_class):
+            for k in keys:
+                if ((not filterset_class) or (k not in filterset_class.base_filters)):
+                    raise ValidationError("invalid filter[{}]".format(k))
+
+        def get_filterset(self, request, queryset, view):
+            """
+            Sometimes there's no filterset_class defined yet the client still
+            requests a filter. Make sure they see an error too. This means
+            we have to get_filterset_kwargs() even if there's no filterset_class.
+
+            TODO: .base_filters vs. .filters attr (not always present)
+            """
+            filterset_class = self.get_filterset_class(view, queryset)
+            kwargs = self.get_filterset_kwargs(request, queryset, view)
+            self.validate_filter(self.filter_keys, filterset_class)
+            if filterset_class is None:
+                return None
+            return filterset_class(**kwargs)
+
+        def get_filterset_kwargs(self, request, queryset, view):
+            """
+            Turns filter[<field>]=<value> into <field>=<value> which is what
+            DjangoFilterBackend expects
+            """
+            self.filter_keys = []
+            # rewrite filter[field] query params to make DjangoFilterBackend work.
+            data = request.query_params.copy()
+            for qp, val in data.items():
+                m = self.filter_regex.match(qp)
+                if m and (not m.groupdict()['assoc'] or
+                          m.groupdict()['ldelim'] != '[' or m.groupdict()['rdelim'] != ']'):
+                    raise ValidationError("invalid filter: {}".format(qp))
+                if m and qp != self.search_param:
+                    if not val:
+                        raise ValidationError("missing {} test value".format(qp))
```

Comment:
not so sure on this validation error. What about when I want to filter for all values with blank string?

Reply:
See [test_filter_empty](https://github.com/django-json-api/django-rest-framework-json-api/pull/466/files#diff-313750acd6d91887834a573075e6dcb8R155) which uses `r'^$'` or you an add a custom filter class as described in [filtering by empty values](https://django-filter.readthedocs.io/en/master/guide/tips.html?highlight=empty#filtering-by-empty-values) and [filtering by an empty string](https://django-filter.readthedocs.io/en/master/guide/tips.html?highlight=empty#filtering-by-an-empty-string) in the django-filter docs.

In django-filter "_empty values are interpreted as a skipped filter._" I thought it better to complain than to return an incorrect result as JSON:API in general follows a pattern of returning errors rather than silently doing the wrong thing. I can see exactly this use case of wanting to filter for an empty string and instead getting no filter at all.

Reply:
I see your point. :+1: